File: app/src/main/java/in/kashewdevelopers/identicon/ControlsActivity.java
package in.kashewdevelopers.identicon;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.SeekBar;

import in.kashewdevelopers.identicon.databinding.ActivityControlsBinding;

public class ControlsActivity extends AppCompatActivity {

    private ActivityControlsBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityControlsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        initialize();
        addBackButton();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    // initialization
    public void initialize() {
        setValuesToWidgets();

        setBorderChangeListener();
        setImageChangeListener();
        setBlockChangeListener();
    }

    public void setValuesToWidgets() {
        // border size 10 - 50 px
        int temp = SharedPrefManager.getBorderSize(this);
        binding.borderSizeController.setProgress(temp - 10);
        binding.borderSizeValue.setText(getResources().getString(R.string.sizePx, temp));

        // image size 150 - 500 px
        temp = SharedPrefManager.getImageSize(this);
        binding.imageSizeController.setProgress(temp - 150);
        binding.imageSizeValue.setText(getResources().getString(R.string.sizePx, temp));

        // no of blocks 3, 5, 7 - 21
        temp = SharedPrefManager.getNoOfBlocks(this);
        binding.noOfBlockController.setProgress((temp - 3) / 2);
        binding.noOfBlockValue.setText(String.valueOf(temp));
    }

    public void addBackButton() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }
    }


    // handle widget clicks
    public void resetToDefaultClicked(View v) {
        binding.borderSizeController.setProgress(SharedPrefManager.DEFAULT_BORDER_SIZE - 10);
        SharedPrefManager.setBorderSize(this, SharedPrefManager.DEFAULT_BORDER_SIZE);

        binding.imageSizeController.setProgress(SharedPrefManager.DEFAULT_IMAGE_SIZE - 150);
        SharedPrefManager.setImageSize(this, SharedPrefManager.DEFAULT_IMAGE_SIZE);

        binding.noOfBlockController.setProgress((SharedPrefManager.DEFAULT_NO_OF_BLOCKS - 3) / 2);
        SharedPrefManager.setNoOfBlocks(this, SharedPrefManager.DEFAULT_NO_OF_BLOCKS);
    }

    public void setBorderChangeListener() {
        binding.borderSizeController.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int borderSize = progress + 10;
                binding.borderSizeValue.setText(getResources().getString(R.string.sizePx, borderSize));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                int borderSize = seekBar.getProgress() + 10;
                SharedPrefManager.setBorderSize(ControlsActivity.this, borderSize);
            }
        });
    }

    public void setImageChangeListener() {
        binding.imageSizeController.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int imageSize = progress + 150;
                binding.imageSizeValue.setText(getResources().getString(R.string.sizePx, imageSize));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                int imageSize = seekBar.getProgress() + 150;
                SharedPrefManager.setImageSize(ControlsActivity.this, imageSize);
            }
        });
    }

    public void setBlockChangeListener() {
        binding.noOfBlockController.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int blocks = 2 * progress + 3;
                binding.noOfBlockValue.setText(String.valueOf(blocks));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                int blocks = 2 * seekBar.getProgress() + 3;
                SharedPrefManager.setNoOfBlocks(ControlsActivity.this, blocks);
            }
        });
    }

}

File: app/src/main/java/in/kashewdevelopers/identicon/ControlsActivity.kt
package `in`.kashewdevelopers.identicon

import `in`.kashewdevelopers.identicon.databinding.ActivityControlsBinding
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.SeekBar
import androidx.appcompat.app.ActionBar

class ControlsActivity : AppCompatActivity() {

    lateinit var binding: ActivityControlsBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityControlsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initialize()
        addBackButton()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }

        return super.onOptionsItemSelected(item)
    }


    // initialization
    private fun initialize() {
        setValuesToWidgets()

        setBorderChangeListener()
        setImageChangeListener()
        setBlockChangeListener()
    }

    private fun setValuesToWidgets() {
        // border size 10 - 50 px
        var temp: Int = SharedPrefManager.getBorderSize(this)
        binding.borderSizeController.progress = temp - 10
        binding.borderSizeValue.text = resources.getString(R.string.sizePx, temp)

        // image size 150 - 500 px
        temp = SharedPrefManager.getImageSize(this)
        binding.imageSizeController.progress = temp - 150
        binding.imageSizeValue.text = resources.getString(R.string.sizePx, temp)

        // no of blocks 3, 5, 7 - 21
        temp = SharedPrefManager.getNoOfBlocks(this)
        binding.noOfBlockController.progress = (temp - 3) / 2
        binding.noOfBlockValue.text = "$temp"
    }

    private fun addBackButton() {
        val actionBar: ActionBar? = supportActionBar
        actionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
            setHomeButtonEnabled(true)
        }
    }

    fun resetToDefaultClicked(view: View) {
        SharedPrefManager.restoreToDefault(this)
        setValuesToWidgets()
    }

    private fun setBorderChangeListener() {
        binding.borderSizeController.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val borderSize = progress + 10
                binding.borderSizeValue.text = resources.getString(R.string.sizePx, borderSize)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                val borderSize: Int
                seekBar?.let {
                    borderSize = it.progress + 10
                    SharedPrefManager.setBorderSize(this@ControlsActivity, borderSize)
                }
            }
        })
    }

    private fun setImageChangeListener() {
        binding.imageSizeController.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val imageSize = progress + 150
                binding.imageSizeValue.text = resources.getString(R.string.sizePx, imageSize)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                val imageSize: Int
                seekBar?.let {
                    imageSize = it.progress + 150
                    SharedPrefManager.setImageSize(this@ControlsActivity, imageSize)
                }
            }
        })
    }

    private fun setBlockChangeListener() {
        binding.noOfBlockController.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val blocks = 2 * progress + 3
                binding.noOfBlockValue.text = "$blocks"
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                val blocks: Int
                seekBar?.let {
                    blocks = 2 * it.progress + 3
                    SharedPrefManager.setNoOfBlocks(this@ControlsActivity, blocks)
                }
            }
        })
    }

}